dot/parachain/availability-store/availability_store_test.go

```diff
@@ -1752,3 +1795,144 @@ func TestStoredDataKeptUntilFinalized(t *testing.T) {
+		ExpectedErasureRoot: availableData2ErasureRoot,
+		Sender:              msgSenderChan,
+	}
+	candidate2Events, err := parachaintypes.NewCandidateEvents()
+	require.NoError(harness.t, err)
+	candidate2Events.Add(parachaintypes.CandidateIncluded{CandidateReceipt: candidate2})
+
+	err = harness.overseer.Start()
+	require.NoError(t, err)
+
+	go harness.processMessages()
+
+	harness.broadcastMessages = append(harness.broadcastMessages, msg1StoreAvailableData)
+	harness.triggerBroadcast()
+
+	// result from seeding data
+	msg1SenderChanResult := <-msg1StoreAvailableData.Sender
+	require.Equal(t, nil, msg1SenderChanResult)
+
+	harness.broadcastMessages = append(harness.broadcastMessages, msg2StoreAvailabeData)
+	harness.triggerBroadcast()
+
+	// result from seeding data 2
+	msgSender2ChanResult := <-msg2StoreAvailabeData.Sender
+	require.Equal(t, nil, msgSender2ChanResult)
+
+	// confirm available data 1 and 2, and has all chunks 1 and 2
+	availableDataResult := harness.queryAvailableData(parachaintypes.CandidateHash{Value: candidate1Hash})
+	require.Equal(t, availableData1, availableDataResult)
+	availableDataResult = harness.queryAvailableData(parachaintypes.CandidateHash{Value: candidate2Hash})
+	require.Equal(t, availableData2, availableDataResult)
+	hasChunks := harness.hasAllChunks(parachaintypes.CandidateHash{Value: candidate1Hash}, nValidators, true)
+	require.True(t, hasChunks)
+	hasChunks = harness.hasAllChunks(parachaintypes.CandidateHash{Value: candidate2Hash}, nValidators, true)
+	require.True(t, hasChunks)
+	harness.printDB("before import leaf")
+
+	// import leaf for candidate 1
+	aLeaf := harness.importLeaf(t, parent1, blockNumber1, candidate1Events)
+	time.Sleep(50 * time.Millisecond)
+
+	// import leaf for candidate 2
+	harness.importLeaf(t, parent2, blockNumber2, candidate2Events)
+	time.Sleep(1500 * time.Millisecond)
+	harness.printDB("after import leaf")
+
+	// signal block 1 finalized for candidate 1
+	blockFinalizedSignal := parachaintypes.BlockFinalizedSignal{
+		Hash:        aLeaf,
+		BlockNumber: blockNumber1,
+	}
+	harness.broadcastMessages = append(harness.broadcastMessages, blockFinalizedSignal)
+	harness.triggerBroadcast()
+	time.Sleep(50 * time.Millisecond)
+
+	// confirm available data 1, and 2, and has all chunks 1 and 2
+	availableDataResult = harness.queryAvailableData(parachaintypes.CandidateHash{Value: candidate1Hash})
+	require.Equal(t, availableData1, availableDataResult)
+	availableDataResult = harness.queryAvailableData(parachaintypes.CandidateHash{Value: candidate2Hash})
+	require.Equal(t, availableData2, availableDataResult)
+	hasChunks = harness.hasAllChunks(parachaintypes.CandidateHash{Value: candidate1Hash}, nValidators, true)
+	require.True(t, hasChunks)
+	hasChunks = harness.hasAllChunks(parachaintypes.CandidateHash{Value: candidate2Hash}, nValidators, true)
+	require.True(t, hasChunks)
+	harness.printDB("after block finalized")
+
+	// wait for pruning
+	time.Sleep(3000 * time.Millisecond)
```

Comment:
does the timing of this consistently work? A little curious about fitting in checks between sleeps, but if it always works that's alright

Reply:
It is working constantly, perhaps it's because it's sort of long, i suspect that if I reduced all the timing by a factor of ten (i.e. time.Sleep(300 * time.millisecond)) that it may cause issues, I'm not sure.